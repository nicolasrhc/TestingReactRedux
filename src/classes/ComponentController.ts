export interface Dato {
    id: number;
    nombre: string;
    apellido: string;
    fecha: Date;
}

class ComponentController {
    public obtenerDatosServidor (): Promise<Dato[]> {
        return new Promise(resolve => (
            resolve([
                {
                    id: 1,
                    nombre: "A",
                    apellido: "_A",
                    fecha: new Date(),
                },
                {
                    id: 2,
                    nombre: "B",
                    apellido: "_B",
                    fecha: new Date(),
                },
                {
                    id: 3,
                    nombre: "C",
                    apellido: "_C",
                    fecha: new Date(),
                }
            ])
        ))
    }
}

export default ComponentController;
