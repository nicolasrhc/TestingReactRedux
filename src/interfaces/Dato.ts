export default interface Dato {
    id: number;
    nombre: string;
    apellido: string;
    fecha: Date;
}