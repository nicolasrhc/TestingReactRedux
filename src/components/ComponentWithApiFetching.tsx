import React, {Component} from "react";
import ComponentController from "../classes/ComponentController";
import Dato from '../interfaces/Dato'

interface ComponentWithApiFetchingState {
    datos: Dato[];
}

interface ComponentWithApiFetchingProps {

}

class ComponentWithApiFetching extends Component<ComponentWithApiFetchingProps, ComponentWithApiFetchingState> {
    static defaultProps: Partial<ComponentWithApiFetchingProps> = {};

    constructor(props: Readonly<ComponentWithApiFetchingProps>) {
        super(props);

        this.state = { datos: [] };
    }

    async componentDidMount(): Promise<void> {
        const componentController: ComponentController = new ComponentController();

        try {
            const datos = await componentController.obtenerDatosServidor();
            
            this.setState({ datos });
        } catch (e) {
            console.error(e);
        }
    }

    render(): JSX.Element {
        return (
            <div>
                <ul>
                {this.state.datos.map((dato, index) => (
                    <li key={dato.id}>{dato.apellido}, {dato.nombre} / {dato.fecha.toDateString()}</li>
                ))}
                </ul>
            </div>
        );
    }
};

export default ComponentWithApiFetching;
