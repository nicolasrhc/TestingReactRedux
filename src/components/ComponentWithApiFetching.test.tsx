import React from "react";
import ComponentWithApiFetching from './ComponentWithApiFetching'
import { shallow, ShallowWrapper } from 'enzyme'
import ComponentController from "../classes/ComponentController";
import Dato from "../interfaces/Dato";

describe("ComponentWithApiFetching Test", () => {
    let wrapper: ShallowWrapper;

    beforeEach(() => {
        
    })

    it("Debe renderizarse correctamente", () => {
        wrapper = shallow(<ComponentWithApiFetching/>);
        expect(wrapper).toMatchSnapshot()
    });

    it("Debe modificar el estado inicial", () => {
        const obtenerDatosServidorMockedResponse: Dato[] = [{id: 1, nombre: "A", apellido: "B", fecha: new Date()}];

        const spyObtenerDatosServidor = jest.spyOn(ComponentController.prototype, "obtenerDatosServidor").mockResolvedValue(obtenerDatosServidorMockedResponse);

        wrapper = shallow(<ComponentWithApiFetching/>);
        
        expect(spyObtenerDatosServidor).toBeCalledTimes(1)
    })
});
