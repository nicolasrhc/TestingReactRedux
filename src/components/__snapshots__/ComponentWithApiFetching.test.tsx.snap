// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentWithApiFetching Test Debe renderizarse correctamente 1`] = `
<div>
  <ul>
    <li
      key="1"
    >
      _A
      , 
      A
       / 
      Wed Jan 29 2020
    </li>
    <li
      key="2"
    >
      _B
      , 
      B
       / 
      Wed Jan 29 2020
    </li>
    <li
      key="3"
    >
      _C
      , 
      C
       / 
      Wed Jan 29 2020
    </li>
  </ul>
</div>
`;
